src/compute/src/render/persist_source.rs

```diff
@@ -0,0 +1,209 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! A source that reads from an a persist shard.
+
+use std::any::Any;
+use std::num::NonZeroUsize;
+use std::rc::Rc;
+use std::sync::Arc;
+use std::task::{Context, Poll};
+use std::time::Duration;
+
+use futures_util::Stream as FuturesStream;
+use mz_storage::source::SourceStatus;
+use timely::dataflow::operators::{Concat, Map, OkErr};
+use timely::dataflow::{Scope, Stream};
+use timely::progress::Antichain;
+
+use mz_dataflow_types::{DataflowError, DecodeError, SourceError, SourceErrorDetails};
+use mz_expr::SourceInstanceId;
+use mz_persist::location::ExternalError;
+use mz_persist_client::read::ListenEvent;
+use mz_persist_client::Location;
+use mz_repr::{Diff, Row, Timestamp};
+
+use crate::compute_state::ComputeState;
+
+/// Creates a new source that reads from a persist shard.
+pub fn persist_source<G>(
+    _render_state: &mut ComputeState,
+    scope: &G,
+    instance_id: SourceInstanceId,
+    consensus_uri: String,
+    blob_uri: String,
+    shard_id: mz_persist_client::ShardId,
+    as_of: Antichain<Timestamp>,
+) -> (
+    Stream<G, (Row, Timestamp, Diff)>,
+    Stream<G, (DataflowError, Timestamp, Diff)>,
+    Rc<dyn Any>,
+)
+where
+    G: Scope<Timestamp = mz_repr::Timestamp>,
+{
+    let worker_index = scope.index();
+
+    let async_stream = async_stream::try_stream! {
+        // We are reading only from worker 0. We can split the work of reading from the snapshot to
+        // multiple workers, but someone has to distribute the splits. Also, in the glorious
+        // STORAGE future, we will use multiple persist shards to back a STORAGE collection. Then,
+        // we can read in parallel, by distributing shard reading amongst workers.
+        if worker_index != 0 {
+            log::trace!("We are not worker 0, exiting...");
+            return;
+        }
```

Comment:
Should the method perhaps instead take a `Vec<ShardId>`, which we can either rig to be non-empty for worker zero only, or to reflect whatever we eventually decide?

Reply:
That's an excellent point! It should be like this, but making the source behave nicely when reading from multiple shards will require some elbow grease. Is it alright if I outline the required changes as a TODO and merge as is?

Reply:
TODOs are great, and idk maybe we should file a general issue to make sure persist sources and sinks are multi-worker. But this should not prevent merging this.